### PR TITLE
Add random and weighted random destroy operators

### PR DESCRIPTION
**Pull request by ekallevik**
Pay closer attention to `get_weighted_probabilities` in `destroy_operators`. I am not quite sure if it is implemented the best way

### CODE REVIEW

File: heuristic/destroy_operators.py [hunk 1]
```diff
@@ -1,14 +1,166 @@
...
+    destroy_set = [
+        set_x(state, t_covered_by_shift, e, t, v, 0)
+        for e in employees
+        for t, v in shifts
+        if state.x[e, t, v] != 0
+    ]
+    return destroy_set
+
+
+def get_weighted_probabilities(score):
```
Comment (ekallevik) [inline heuristic/destroy_operators.py:152]: This is something we should discuss on Monday

File: heuristic/alns.py [hunk 2]
```diff
@@ -76,15 +75,61 @@ def __init__(self, state, data, criterion):
             worst_week_removal,
             self.competencies,
             self.time_periods_in_week,
-            self.combined_time_periods_in_week,
```
Comment (ekallevik) [inline heuristic/alns.py:79]: This was an unused argument

File: heuristic/alns.py [hunk 3]
```diff
@@ -76,15 +75,61 @@ def __init__(self, state, data, criterion):
...
+            self.employees,
+            self.weeks,
+            self.shifts_at_day,
+            self.t_covered_by_shift,
+            self.random_state,
+        )
+
         remove_worst_employee = partial(
-            worst_employee_removal, self.shifts, self.t_covered_by_shift, self.competencies
+            worst_employee_removal, self.shifts, self.t_covered_by_shift,
```
Comment (ekallevik) [inline heuristic/alns.py:116]: Same goes for this. Competencies is an unused argument

File: heuristic/alns.py [hunk 4]
```diff
@@ -83,8 +85,63 @@ def __init__(self, state, data, criterion):
...
+        remove_weighted_random_employee = partial(
+            weighted_random_employee_removal,
+            self.shifts,
+            self.t_covered_by_shift,
+            self.competencies,
+            self.employees,
+            self.random_state,
         )
 
         repair_worst_week_regret = partial(
```
Comment (hkgrov) [inline heuristic/alns.py:147]: Some sets are missing here that are included in daily_rest. Must be sure they are not overwritten with this merge.

File: heuristic/alns.py [hunk 5]
```diff
@@ -293,31 +359,16 @@ def initialize_random_state():
...
-        for new_operator in repair_operators:
-            self.repair_operators[destroy_operator_id][new_operator.func.__name__] = new_operator
-
     def add_destroy_and_repair_operators(self, operators):
-        for key in operators.keys():
-            self.add_destroy_operator([key])
-            self.add_repair_operator(key.func.__name__, operators[key])
 
-    @staticmethod
-    def add_operator(operators, new_operator):
```
Comment (hkgrov) [inline heuristic/alns.py:310]: Have you testet and are sure that all operators are included and that they are run?

**Comment by hkgrov** [comment]
Overall it looks good. I am mostly concerned with not overwriting differences introduced in alns_daily_rest as they would most likely not result in error codes, but would give wrong answers nevertheless.

File: heuristic/alns.py (hunk 5 above)
Comment (ekallevik) [reply heuristic/alns.py:310]: Yes!

File: main.py [hunk 6]
```diff
@@ -86,35 +86,57 @@ def set_alns(self):
...
+                for c in self.data["competencies"]
+                for t in self.data["time"]["periods"][0]
+            },
+            "daily_rest_error": {
+                (e, i): 0
+                for e in self.data["staff"]["employees"]
+                for i in self.data["time"]["days"]},
+
             "delta_positive_contracted_hours": {e: 0 for e in self.data["staff"]["employees"]},
-            "daily_rest_error": {(e,i): 0 for e in self.data["staff"]["employees"] for i in self.data["time"]["days"]}
```
Comment (hkgrov) [inline main.py:113]: Daily_rest error må inkluderes igjen.

**Pull request merged**